Kronuz/cpython
# Check imports eagerly default

# Summary:
Add a test to check if Python imports are executed eagerly in default.

# Test Plan:
First of all, we need to check if it is in default (Lazy Imports is not enabled). If Lazy Imports is enabled, we do nothing in this test. On the other hand, if it is in default, we should check if imports are executed eagerly. Thus, this test is divided into two sections.
1. Check Lazy Imports is enabled or not
2. Check imports are executed eagerly in default

We create two classes to complete these two points. `LazyImportsEnabledChecker` is for the section one, and `TestEagerlyImport` is for the section two. Moreover, this test is wrapped by `unittest`, so we could just check the result and know if it works correctly.

## Check Lazy Imports is enabled or not
Two options make Lazy Imports work, and we check both of them in this test.
### 1. `-L` flag

If we add the `-L` or `-l` to our execution command, it should show _success_.
e.g.
```
./python.exe -mtest -v test_eagerly_import -L 
```
Expected result:
<img width="211" alt="image" src="https://user-images.githubusercontent.com/18440101/172500582-0bc32da6-4b2e-404b-a6b1-1291d1aa5671.png">

```
./python.exe -mtest -v test_eagerly_import -l
```
Expected result:
<img width="209" alt="image" src="https://user-images.githubusercontent.com/18440101/172500623-21e39aa5-7d4d-4f77-8232-b539909042d2.png">


### 2. `PYTHONLAZYIMPORTSALL` environment variable
First, run this to set up the environment variable of Lazy Imports.
```
export PYTHONLAZYIMPORTSALL=1
```
Then, run this to check if it succeeds in without the flag.
```
./python.exe -mtest -v test_eagerly_import 
```
Expected result:
<img width="213" alt="image" src="https://user-images.githubusercontent.com/18440101/172501147-6badd46d-8da2-46f7-9284-afad0ce9aa42.png">


## Check imports are executed eagerly in default
We use `json` and `sqlite3` to confirm it. 
### json
`json` is not imported in the test code, so it and its sub-modules should not be loaded into `sys.modules`.
Thus, we check its sub-modules `json.encoder` and `json.decoder` in unit test to know if they are imported.

### sqlite3
After excluding all the edge cases, we could start to test if imports are really loaded when they are imported. We imported `sqlite3` at the beginning of the code (line 15), so its sub-module `sqlite3.dbapi2` should be in `sys.modules` even if we do not use it yet.

### Test behaviors in default
Open a new terminal (without setting `PYTHONLAZYIMPORTSALL`) and run 
```
./python.exe -mtest -v test_eagerly_import
```
Expected result:
<img width="203" alt="image" src="https://user-images.githubusercontent.com/18440101/172503077-dda9667e-cef0-4c3e-9a6d-125f98fa0db9.png">
